### PR TITLE
Spring boot practice 2

### PR DESCRIPTION
**Pull request by VasylAzarov**
Implemented cat logic with entities, service, repository,mapper, controller, dtos | implemented endpoints for getting cart, adding to cart, updating cart, deleting from cart

### CODE REVIEW

File: src/main/java/dev/vasyl/proj/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    public CartResponseDto getCart(@AuthenticationPrincipal User user) {
+        return cartManagementService.getCart(user);
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    @Operation(summary = "Add item to user cart",
+            description = "Add item to user cart. Available for User")
+    public CartItemDto addItem(@AuthenticationPrincipal User user,
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/controller/ShoppingCartController.java]:
```suggestion
    public CartResponseDto addItem(@AuthenticationPrincipal User user,
```

File: src/main/java/dev/vasyl/proj/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                               @RequestBody
+                               @Valid CreateCartItemRequestDto createCartItemRequestDto) {
+        return cartManagementService.save(user, createCartItemRequestDto);
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @PutMapping("/items/{cartItemId}")
+    @Operation(summary = "Update item quantity in user cart",
+            description = "Update item quantity in user cart. Available for User")
+    public CartItemDto updateItem(@PathVariable Long cartItemId,
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/controller/ShoppingCartController.java]:
```suggestion
    public CartResponseDto updateItem(@PathVariable Long cartItemId,
```

File: src/main/java/dev/vasyl/proj/dto/shopping/cart/CreateCartItemRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import jakarta.validation.constraints.NotNull;
+
+public record CreateCartItemRequestDto(
+        @Schema(description = "Book id",
+                example = "2")
+        @NotNull(message = "id is empty")
+        Long bookId,
+        @Schema(description = "Cart item quantity",
+                example = "1")
+        @Min(0)
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/dto/shopping/cart/CreateCartItemRequestDto.java]:
```suggestion
        @Positive
```

File: src/main/java/dev/vasyl/proj/dto/shopping/cart/UpdateCartItemRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import io.swagger.v3.oas.annotations.media.Schema;
+import jakarta.validation.constraints.Min;
+
+public record UpdateCartItemRequestDto(
+        @Schema(description = "Operation increase(be default) quantity or decrease quantity",
+                example = "INCREASE")
+        CartItemOperation operation,
+        @Schema(description = "quantity",
+                example = "3")
+        @Min(0)
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/dto/shopping/cart/UpdateCartItemRequestDto.java]:
```suggestion
        @Positive
```

File: src/main/java/dev/vasyl/proj/model/CartItem.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+@Entity
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/dev/vasyl/proj/model/CartItem.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/dev/vasyl/proj/model/ShoppingCart.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+@Entity
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/model/ShoppingCart.java]: you can use `@MapsId` here

File: src/main/java/dev/vasyl/proj/service/impl/CartManagementServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+import dev.vasyl.proj.service.CartManagementService;
+import java.util.Set;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CartManagementServiceImpl implements CartManagementService {
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/service/impl/CartManagementServiceImpl.java]:
```suggestion
public class ShoppingCartServiceImpl implements CartManagementService {
```

File: src/main/java/dev/vasyl/proj/service/CartManagementService.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+import dev.vasyl.proj.dto.shopping.cart.CartItemDto;
+import dev.vasyl.proj.dto.shopping.cart.CartResponseDto;
+import dev.vasyl.proj.dto.shopping.cart.CreateCartItemRequestDto;
+import dev.vasyl.proj.dto.shopping.cart.UpdateCartItemRequestDto;
+import dev.vasyl.proj.model.User;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+
+public interface CartManagementService {
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/service/CartManagementService.java]:
```suggestion
public interface ShoppingCartService {
```

File: src/main/java/dev/vasyl/proj/service/impl/CartManagementServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    private final BookRepository bookRepository;
+    private final CartManagementMapper cartManagementMapper;
+
+    @Override
+    public CartResponseDto getCart(User user) {
+        return cartManagementMapper.toCartDto(getShoppingCartByUserId(user.getId()));
+    }
+
+    @Override
+    public Page<CartItemDto> findAll(User user, Pageable pageable) {
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/service/impl/CartManagementServiceImpl.java]:
```suggestion
    public Page< CartResponseDto > findAll(User user, Pageable pageable) {
```

File: src/main/java/dev/vasyl/proj/service/impl/CartManagementServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+import dev.vasyl.proj.repository.CartItemRepository;
+import dev.vasyl.proj.repository.ShoppingCartRepository;
+import dev.vasyl.proj.service.CartManagementService;
+import java.util.Set;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+@Service
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/service/impl/CartManagementServiceImpl.java]: Think where is `@Transactional` needed

File: src/main/java/dev/vasyl/proj/service/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -32,12 +35,26 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
...
         Role role = roleRepository.findByName(RoleName.USER).orElseThrow(
                 () -> new EntityNotFoundException("Error when set user role"));
         user.setRoles(Set.of(role));
         userRepository.save(user);
-        return userMapper.toUserResponse(user);
+        createShoppingCartForUser(user);
+        return user;
+    }
+
+    private void createShoppingCartForUser(User user) {
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/service/impl/UserServiceImpl.java]: replace this method to ShopCartService

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 13]
```diff
@@ -17,3 +17,7 @@ databaseChangeLog:
       file: db/changelog/changes/08-create-categories-table.yaml
   - include:
       file: db/changelog/changes/09-create-books_categories-table.yaml
+  - include:
+      file: db/changelog/changes/10-create-shoping_carts-table.yaml
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/db.changelog-master.yaml]:
```suggestion
      file: db/changelog/changes/10-create-shopping_carts-table.yaml
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/dev/vasyl/proj/mapper/ShoppingCartMapper.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    CartResponseDto toCartDto(ShoppingCart shoppingCartByUserId);
+
+    @Named("cartItemsSetToFirstPageCartItemDto")
+    default Page<CartItemResponseDto> cartItemsSetToFirstPageCartItemDto(Set<CartItem> cartItems) {
+        List<CartItemResponseDto> dtoList = cartItems.stream()
+                .map(this::toCartItemDto)
+                .toList();
+        return dtoList.isEmpty()
+                ? Page.empty()
+                : new PageImpl<>(dtoList, PageRequest.of(0, dtoList.size()), dtoList.size());
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/mapper/ShoppingCartMapper.java]: you can use the method map of Page and simplify this method

File: src/main/java/dev/vasyl/proj/mapper/ShoppingCartMapper.java [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Named("toCartItemsDtoPage")
+    default Page<CartItemResponseDto> toCartItemsDtoPage(Page<CartItem> cartItemPage) {
+        List<CartItemResponseDto> dtoList = cartItemPage.getContent().stream()
+                .map(this::toCartItemDto)
+                .toList();
+        return new PageImpl<>(dtoList, cartItemPage.getPageable(),
+                cartItemPage.getTotalElements());
+    }
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/mapper/ShoppingCartMapper.java]: the same

File: src/main/java/dev/vasyl/proj/repository/ShoppingCartRepository.java [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+import org.springframework.data.jpa.repository.EntityGraph;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+
+    @EntityGraph(attributePaths = {"cartItems", "cartItems.book"})
+    Optional<ShoppingCart> findByUserId(Long userId);
+
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/repository/ShoppingCartRepository.java:14]:
```suggestion
    @EntityGraph(attributePaths = {"cartItems", "cartItems.book"})
    Optional<ShoppingCart> findByUserId(Long userId);
```

File: src/main/java/dev/vasyl/proj/service/impl/ShoppingCartServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+    private Book getBookById(Long bookId) {
+        return bookRepository.findById(bookId).orElseThrow(
+                () -> new EntityNotFoundException(
+                        "Can`t find Book entity by book id: " + bookId));
+    }
+
+    private CartItem getCartItemIfExist(ShoppingCart shoppingCart, Book book) {
+        Long bookId = book.getId();
+        Set<CartItem> cartItemSet = shoppingCart.getCartItems();
+        return cartItemSet.stream()
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/service/impl/ShoppingCartServiceImpl.java:97]: better return Optional instead  of null and use Optional methods for conditional

File: src/main/java/dev/vasyl/proj/service/impl/ShoppingCartServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    private int getCalculatedQuantity(CartItem cartItem,
+                                      UpdateCartItemRequestDto updateCartItemDto) {
+        int currentQuantity = cartItem.getQuantity();
+        int change = updateCartItemDto.quantity();
+        if (updateCartItemDto.operation() == CartItemOperation.DECREASE
+                && currentQuantity < change) {
+            return 0;
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/service/impl/ShoppingCartServiceImpl.java]: Better throw some Exception and return 4xx response

File: src/main/java/dev/vasyl/proj/service/impl/ShoppingCartServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+        return updateCartItemDto.operation() == CartItemOperation.DECREASE
+                ? currentQuantity - change
+                : currentQuantity + change;
+    }
+
+    public void createShoppingCartForUser(User user) {
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/service/impl/ShoppingCartServiceImpl.java]: move it before private methods. Override annotation missed

File: src/main/java/dev/vasyl/proj/service/impl/UserServiceImpl.java [hunk 20]
```diff
@@ -32,12 +34,20 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
...
                     + "] already exist");
         }
-        User user = userMapper.toModel(requestDto);
-        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+
+        User user = saveUser(userMapper.toModel(requestDto),
+                passwordEncoder.encode(requestDto.getPassword()));
+
+        return userMapper.toUserResponse(user);
+    }
```
Comment (liliia-ponomarenko) [inline src/main/java/dev/vasyl/proj/service/impl/UserServiceImpl.java:44]: Should you use Transactional here?

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 21]
```diff
@@ -16,4 +16,8 @@ databaseChangeLog:
   - include:
       file: db/changelog/changes/08-create-categories-table.yaml
   - include:
-      file: db/changelog/changes/09-create-books_categories-table.yaml
+      file: db/changelog/changes/09-create-books-categories-table.yaml
+  - include:
+      file: db/changelog/changes/10-create-shoping-carts-table.yaml
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/db.changelog-master.yaml]:
```suggestion
      file: db/changelog/changes/10-create-shopping-carts-table.yaml
```

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  fix a few comments ;)

File: src/main/java/dev/vasyl/proj/controller/ShoppingCartController.java [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+    public CartResponseDto getCart(@AuthenticationPrincipal User user) {
+        return shoppingCartService.getCart(user);
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    @Operation(summary = "Add item to user cart",
+            description = "Add item to user cart. Available for User")
+    public CartItemResponseDto addItem(@AuthenticationPrincipal User user,
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/controller/ShoppingCartController.java]:
```suggestion
    public CartResponseDto addItem(@AuthenticationPrincipal User user,
```

File: src/main/java/dev/vasyl/proj/controller/ShoppingCartController.java [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+                                       @RequestBody
+                               @Valid CreateCartItemRequestDto createCartItemRequestDto) {
+        return shoppingCartService.save(user, createCartItemRequestDto);
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @PutMapping("/items/{cartItemId}")
+    @Operation(summary = "Update item quantity in user cart",
+            description = "Update item quantity in user cart. Available for User")
+    public CartItemResponseDto updateItem(@PathVariable Long cartItemId,
```
Comment (Elena-Bruyako) [inline src/main/java/dev/vasyl/proj/controller/ShoppingCartController.java]:
```suggestion
    public CartResponseDto updateItem(@PathVariable Long cartItemId,
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/dev/vasyl/proj/model/ShoppingCart.java [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+@Setter
+@Entity
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    private Long id;
+
+    @OneToOne
```
Comment (JJJazl) [inline src/main/java/dev/vasyl/proj/model/ShoppingCart.java]:
```suggestion
    @OneToOne(fetch = FetchType.LAZY)
```

File: src/main/java/dev/vasyl/proj/service/impl/ShoppingCartServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,117 @@
...
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final BookRepository bookRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+
+    @Override
+    public CartResponseDto getCart(User user) {
+        return shoppingCartMapper.toCartDto(getShoppingCartByUserId(user.getId()));
+    }
+
+    @Transactional
```
Comment (JJJazl) [inline src/main/java/dev/vasyl/proj/service/impl/ShoppingCartServiceImpl.java]: You can move `@Transactional` above the class
Maybe this article will help you better understand how these annotations and pattern Proxy work:
https://medium.com/@denis.verkhovsky/spring-how-annotations-work-part-2-f9ea764abc28

File: src/main/java/dev/vasyl/proj/service/impl/UserServiceImpl.java [hunk 26]
```diff
@@ -32,12 +35,21 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
                     + requestDto.getEmail()
                     + "] already exist");
         }
-        User user = userMapper.toModel(requestDto);
-        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+
+        User user = saveUser(userMapper.toModel(requestDto),
```
Comment (JJJazl) [inline src/main/java/dev/vasyl/proj/service/impl/UserServiceImpl.java:40]: Add `@Transactional` above the class

**Review by JJJazl — APPROVED** [review]

**Pull request merged**